src/features/messages/cards/TransactionCard.tsx

```diff
@@ -236,11 +249,59 @@ function CallDataModal({ debugResult }: { debugResult?: MessageDebugResult }) {
           </p>
           <LabelAndCodeBlock label="Recipient contract address:" value={contract} />
           <LabelAndCodeBlock label="Handle function input calldata:" value={handleCalldata} />
+          <button onClick={handleClick}
+            disabled={loading}
+            className='underline text-blue-400'
+          >
+            {buttonText}
+          </button>
+          {loading && <Spinner classes="mt-4 scale-75 self-center" />}
         </div>
       </Modal>
     </>
   );
 }
+async function simulateCall({contract,handleCalldata,chainId,message}:{contract:string,handleCalldata:string,chainId:ChainId,message:Message}){
+
+  
+  const data:SimulateBody={
+        save: true,
+        save_if_fails: true, 
+        simulation_type: 'full',
+        network_id: chainId, 
+        from: '0x0000000000000000000000000000000000000000',//can be any address, doesn't matter
+        to: contract,
+        input:handleCalldata,
+        gas: BigNumEth.from(message.totalGasAmount).toNumber(),
+        gas_price: Number(computeAvgGasPrice("wei",message.totalGasAmount,message.totalPayment)),
+        value: 0,
+  }
+  const resp=await fetch(
+    `/api/simulation`,{
+      method:'POST',
+      body:JSON.stringify(data),
+    }
+  )
+  
+  const simulationId=await resp.json().then((data)=>data.data)
+  window.open(`https://dashboard.tenderly.co/shared/simulation/${simulationId}`)
+  
+}
+
+function computeAvgGasPrice(unit: string, gasAmount?: BigNumber.Value, payment?: BigNumber.Value) {
```

Comment:
Why did you duplicate this function here? Please de-dupe with the copy in GasDetailsCard and move it to `messages/utils.ts`